### DISCUSSION TITLE
Reworking EasyOCR

### DISCUSSION THREAD

**Discussion by JulianOrteil** [JaidedAI/EasyOCR#839] [Ideas]
This discussion relates to reworking EasyOCR and bringing it up to standards set by various PEPs and any suggested here.

**The issue:**
EasyOCR is a very popular library within the machine vision community. It competes with top-of-the-line software like Tesseract while maintaining a robust community. However its many achievements--which shouldn't be understated--the library falls short in multiple areas like documentation, testing, and readability; among others.

While perhaps not an issue for large swathes of users, these hindrances do present issues to others who want to contribute, fine-tune, or just learn OCR in general. Because of this, @ystoll and I have decided to donate our time to overhauling EasyOCR to address all of these issues while maintaining its excellent performance and ease of use for the typical user.

**Breakdown:**
Information about specific pain points we've identified is spread between two issues (#823 and #829); they'll be consolidated below. This is by no means an exhaustive list, so if you have a suggestion, please feel free to share it in a reply.

- EasyOCR is severely non-PEP compliant. Some specifics include variable names not being `snake_case`, code comments being strings (not `# comment`), unused variables, etc. All of these severely reduce the readability of EasyOCR and do present themselves as a hurdle for potential contributors.
- Documentation is severely lacking. Especially related to how parameters affect detection in methods like `Reader.detect`. The library presents itself as "ready-to-go, out-of-the-box" and the current documentation may be sufficient for less-adept users. However, "power users" will spend considerable amounts of time tweaking parameters to find the optimal outcome for their problems and should be able to discern what they do from reading the docstrings. Not to mention, the documentation hosted on JaidedAI's website is separate from the code. Using ReadTheDocs, documentation can be updated with code commits automatically, reducing overhead.
- Type hints are virtually non-existent. These are extremely useful for static type checkers and users to determine what possible types of values a parameter expects.
- Tests don't exist. The library appears to be largely tested through usage which isn't proper. These are starting to be added through https://github.com/JaidedAI/EasyOCR/commit/9bd8be01cb83dded4cf41af9deb3f582539c0a5d, but they are being added in a way that emphasizes the previous point about readability.
- Everything is manual. The library is not automated in any real way for performing tasks such as running tests, producing releases, etc. These can be easily performed through tools like tox and makes contributors' lives much easier.

**Approach:**
@rkcosmos stated in #823 that he is concerned about one massive PR being used for a project like this. This is a valid point. To appease this concern, the best approach will likely be making changes, with compatibility, over multiple versions. This allows for a more seamless change in the code base--keeping the old API for a couple of versions with deprecation warnings alongside the new API--but does introduce much compatibility work on top of the proposed changes.

**Plan:**
For this rework to take place, a game plan should be determined before any work is performed. As such, I propose the following:
1. The first thing to tackle should be automating tests, releases, linting, etc (anything related to CI/CD as well). This is typically in the form of `tox` and/or GitHub Actions and is the least invasive change. Because it doesn't affect the main source code, this change will be invisible to end-users. This also sets up workflows for further steps.
2. Tests, tests, tests. Before any changes are made to the library, we need to add tests to establish the "check and balance" on our changes. If our changes break the tests, then those changes shouldn't be accepted.
3. Next should be type-hints. Doing this as the first code change kills three birds with one stone: provides type hints, allows us to get a better handle on how data flows throughout the code, and helps automate documentation.
4. Start documenting. Like type-hints, this takes care of multiple issues: documentation itself and further understanding of what each function in the API is doing.
5. Identify and eliminate things discussed in the points above relating to PEP compliance--to an extent. We only want to put in as much work as necessary to make reworking easier. Optimizations should not happen, and the cleaning of code should be kept to a minimum so as to reduce unnecessary overhead.
6. Start implementing the new architecture. However, no code should be moved from old to new, the new should just call the old like a proxy. This allows us to start adding `DeprecationWarning`s everywhere in the old code telling users to start using the new API while maintaining the old API as-is. Our changes should be invisible to the end user unless they absolutely cannot be avoided.
7. Systematically start porting code. This is where things will get muddy because we need to keep old code for at least 2 minor revisions. Also, new tests, documentation, and other necessary functions should be performed at this time. This is where the hard work will be as this step should be broken down into multiple PRs addressing specific modules or classes.

Step 7 should be the extent of this rework. Obviously, as we work through these changes, there may be areas where we can improve or optimize code, but that should be left for another day. The entire purpose of this rework should be limited to just addressing the above points and any points discussed below.

If you have any ideas, concerns, questions, or other comments related to this rework you feel would be beneficial to add, then please post them. We are also open to adding additional contributors to this project if you feel the desire to.

**Comment by ystoll** [comment]
@JulianOrteil  Thank you for summarizing the TODO list !
One concern though, #839 does not seem to appear in the issue list, is this normal ?
I will start to work on a PyTest tests suit myself and will propose a PR when I will have completed a few of them.
Please note that I am not an expert in unitary testing (I have a physics background), although I will do my best, remarks on my contributions will be more than appreciated. 

@rkcosmos You said that you were concerned about the stability of the package w.r.t PR we can make: maybe you can create a new development branch into which we will merge our PRs. You can then test intensively our changes before merging it into the master branch ?

**Reply by JulianOrteil** [reply]
@ystoll Yes, that is correct. This is a discussion, not an issue, so you can find this under the `Discussions` tab above. I feel putting our thoughts and progress updates here is more appropriate than an issue.

Before you start, please add me as a collaborator on your fork. I can start working on adding workflows to help you with automating testing.

**Reply by ystoll** [reply]
@JulianOrteil  Yes, I will do that. Thank you for your response.

**Reply by JulianOrteil** [reply]
@ystoll Perfect, thank you. Let me know when you get me added and I'll hammer it out quickly.

**Reply by ystoll** [reply]
@JulianOrteil 

Hi Jullian,
you have been added as a collaborator to my private fork.
A few remarks:
you will find three branches in this fork:
1) Master branch: This branch will be kept sync to the official repo master branch.
2) tests_suit: this is the branch where I intend to create the Pytest based test suit. There is not much for now in it, but nonetheless, this is where I propose we develop the suit.
3) refactoring_pep_docstring: this is where I have started to make the code PEP compliant:

- I have used the black auto-formatter as a starting point to correctly align and justify the code.
    Just a thing about black, it tends to format functions definitions as follow:
     ```
         def my_func(
                     a, 
                     b
               ):
    ```
It does so, in order to respect the maximum number of characters per line.
I tend to prefer to format functions and methods definitions like so:
 ```
def my_func(a,
             b):
 ```
even though it means that we have to forget about the 72 characters rule per line.
Of course, we can discuss this but, personally, I find the latter easier to read (a.k.a we see in an wink where the list of parameters starts
and where the list of parameters ends).

- On this branch, I have started to snake_case all variables that I could. I also suppressed most of unused variables and solve a few other stylistic issues (for instance, I use immutable tuples instead of lists for default arguments, or I have added a verbose option in certain functions in order to display Errors messages that were just dismissed until then).

- Certain variables names starts with "curr_" as in "curr_norm_img", here "curr_" stands for "current_", this is a personal convention
to reminds me that this is the object/value that is indeed used at the time that the function/method is called.

- I have started to introduce docstrings template in certain functions. We should agree on a standard format for all docstrings.
    Here again, personally, I prefer to use the numpy standard [numpy_docstrings](https://numpydoc.readthedocs.io/en/latest/format.html) for readability but of course, we can discuss this.

- There is one variable name exposed to the public API that I have changed: I have snake_case beamWidth to beam_width. I am aware that this one will need to be restored before we make any PR.

- requirements_perso.txt list all the packages I need to create a venv that fits my hardware (GPU and CUDA versions) and that are compatible with the package.

To the best of my knowledge, the modifications I made on this branch did not introduce any regressions.
Of course, until such times where we will have a proper tests suit, there will be no way to be 100% sure that I did not messed up anything.
**I am aware that the changes I made up in this branch will certainly need to be split up in smaller chunks to be acceptable by @rkcosmos**.

Please let me know if there is something else unclear,
cordially

Yannick

**Reply by JulianOrteil** [reply]
@ystoll Thanks for the invite. I agree, let's hash out some conventions before we begin. I'll do my best to address each of your bullets with a list of my own.

- I suggest the 72-character limit be pushed up to 100 for code. I find, in my practice, that there are far too many instances where 72 characters are not enough. However, I usually keep the 72-character limit for docstrings and comments (excluding links and Sphinx references, among others).

- For the formatting of functions, I also use black (within pre-commit). Is the format you use something configurable with black as I typically keep most settings in it as default? If not, this will be a no-go from me since I also rely on pre-commit (explained below). Also, in your experience, do you ever see your type-hints exceed the character limit and/or become too complex for the line when using your preferred format?

- I am perfectly fine with your variable name conventions. As long as the acronyms are common vernacular, you'll get no complaints from me.

- As for docstrings, I would prefer Google-style as it seems to be the most readable to me; however, I am not very partial to any specific type. Just so long as Sphinx can parse it and generate documentation for us.

- I won't touch your requirements file; although, I recommend you look into Miniconda for AI-GPU related development. It might make setting up and tearing down environments easier if you find it to be a hassle currently.

Now onto my suggestions:
- As I stated, I usually develop under the pre-commit pipeline. In other words, when I try to commit code, this handy utility runs which sends my changes through tools such as mypy, flake8, bandit, black, isort, etc. If anything fails these linters, the commit fails. I view this tool as a necessity for keeping consistency between collaborators, so we should really consider adding it to our cycle. (If black continues to "correct" your style of function, then commits will always fail, hence my no-go if your style is not configurable).

- Going off of the above point, I also use a super strict linter called `wemake-python-styleguide`. It's a massive linter combing pylint, black, bandit, and many others; into one single cohesive tool. But, as I said, it is super strict and will flag literally everything wrong with what you do. I love it and will use it, but I won't force you to if you don't want to.

- I'll be adding automation tools such as `tox` very soon and setting them up. As EasyOCR is catered towards multiple different versions of Python, I highly suggest you install Miniconda. `tox` can automatically run tests across any major Python version in its own dedicated conda environment.

Once I have tox setup, I'll give you a quick introduction to it if you would like it. I'll also set up my own branch for library metadata and automation.

My best,
Jules

**Reply by JulianOrteil** [reply]
@ystoll tox has been added with additional files. I can look into adding Github Actions if this is something @rkcosmos would like to see added?

**Reply by ystoll** [reply]
Hi @JulianOrteil,

Thank you for your response.
Generals thoughts:
- I went through tutorials for pre-commit and tox, and these seem like great tools! 
- I wasn't that much familiarize with these tools, but it seems like the perfect occasion to learn ! Turns out I was searching a pretext for developing skills in packaging and CI/CD in python, thank you for giving me one !
- The pipeline you propose seems really neat, with very high standards. And that is great.
- However, given the codebase we are starting with, I would like to suggest that we set constraints incrementally:
   - Indeed, I was trying to use precommit but with the current code being full of inconsistencies w.r.t PEP, it is a nightmare to do so.
   - Linters returns specific warnings that, if corrected, would imply to rewrite entirely certain functions like warnings complaining about functions being to complex cognitively (warnings returned by Flake) and in first steps of the refactoring, we shouldn't do that.
   - As you and @rkcosmos said we should make relatively atomic changes to start with in order for our modifications to be accepted in the main repository.

 My suggestions:

- In a first step, we should turn off most off pre-commit hooks, keeping in mind that we will turn them back on, when we will reach a cleaner codebase.
- We should really focusing on the test suits, however concerning the tests, I had a question. Do you think that we should either:
  1. Leave a badly formatted function untouched and write the test accordingly:
         ```def badFunction(a, b): return a+b, def test_badFunction(): a = 1, b = 2 assert badFunction(a, b) == 3```
  2.  Or modify the function name and write a proper test to it:
          ```def good_function(a, b): return a+b, def test_good_function(): a = 1, b = 2 assert good_function(a, b) == 3```
P.S: sorry, but I met difficulties introducing linebreaks in codeblocks inside nested lists in GitHub markdown, 
see [issue_md](https://github.com/CenterForOpenScience/osf.io/issues/7308)

  Concerning your thoughts:
 - Black: I thought about it, and yeah, let's keep things simple and stay with the default in Black.
 - Docstrings: It seems that both Google and Numpy standards can be treated by Sphynx [docstrings_sphynx](https://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html).
 - That being said, I do not like extra parenthesis used in Google docstrings for type annotations. Hence I would be more at ease with numpy docstyles which are more or less what is used by the Scikit Learn ecosystem [doc_np](https://scikit-beam.github.io/scikit-beam/development/docrules.html) that I tend to mimic.

 Conclusion:
 For starting, as a "kick-off" phase:
 1. We should decide which constraints we keep on during our "kick-off" phase, which ones we *temporally* dismiss. 
 2. We should propose a first PR with unused variables removed and black formatted code. 
 3. Tests, tests, tests: we should propose a proper tests suits.
 4. and write docstrings, at least for the mains functions (the documentation coverage could be improved later).

Do you agree with what I am suggesting ?

F.Y.I, I will be on vacations for the next two weeks, so I will be offline during this period.

Cordially

Yannick

**Reply by JulianOrteil** [reply]
@ystoll

- I agree, pre-commit is a bit much at the moment. It can be limited to just the files you're working on, but is not able to limit to just your changes in those files. I can see what linters should be disabled, such as flake8, but leave others like mypy and isort enabled. The other thing to consider is that I have set up a tox environment (run `tox -e lint`) to run pre-commit for us. I think right now, using that instead of having pre-commit properly enabled will save us a bunch of headaches.

SIDE NOTE: If you want to see what environments are available in the tox config I created, run `tox -av` in your terminal.

- For tests, the intention of my proposal was to test the codebase as-is without touching any source code beforehand. This is so we can establish a baseline for our future changes to be tested against. So, to answer your question, don't change any function names. If anything, avoid touching the source code at all until tests are done.

- For docstrings, Google-style does not require parenthesis for specifying argument types if they are annotated. I recommend you look at [this resource](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) for some examples of what I mean (specifically `function_with_pep484_type_annotations`).

- Another note about documentation, there are three main styles of online documentation we can use: furo, alabaster, and rtd_theme. There are many more, but I suggest sticking to one of these three as they are the most popular and familiar to most users. Examples of all three can be found here: [KivyMD (furo)](https://kivymd.readthedocs.io/en/latest/), [wemake-python-styleguide (alabaster)](https://wemake-python-styleguide.readthedocs.io/en/latest/), [tda-api (rtd_theme)](https://tda-api.readthedocs.io/en/latest/). Personally, I like furo the best as I think it to be the cleanest and easiest to navigate, but I am comfortable using any of the three and leave the decision to you and/or @rkcosmos.

- To finalize, I agree with your schedule proposal but would like to see 2 and 3 switched. This is just so we can avoid accidentally removing things and not catching them because we don't have any tests.

And no worries, I hope you have a fun vacation! I'll get working on laying some of the groundwork on my end and start writing tests while you're away.

Have fun!
Jules

**Reply by JulianOrteil** [reply]
@ystoll I believe I have finished enough automation for the development process that we should be ready to merge the `metamation` branch. I can submit a PR, but your input would be beneficial on anything I missed or you would like to see changed beforehand.

Of note, due to the complexity of the build process EasyOCR currently employs, `conda` is required for it to complete. Please note the following changes and notes about my work:
- Fully building the library is only possible on Linux. I have not yet worked on trying to get it to complete on Windows, so that may be my next task but building on Windows will likely result in DBNet failing as it currently stands.
- For reasons I am not fully able to diagnose, `python setup.py install` and `python setup.py develop` fail to compile DBNet. I believe this is because these commands spawn a temporary build environment that does not contain all components required for the successful building of DBNet. I am not sure if this is intended, but it fails nonetheless. Because of this, I have replaced the `cmdclass` instructions.
- The `cmdclass` instructions have been replaced by `tox -e build`. This instruction should handle everything necessary to compile EasyOCR in its entirety, including DBNet. I highly recommend you run this instruction before `pip install -e .` or `pip install .` As an added bonus, due to `tox -e build`, DBNet should now be able to ship with base EasyOCR, or as an extra (i.e. `pip install easyocr[dbnet]`). We can explore this option later.
- A new tox instruction, `tox -e cleanall` has been added. This instruction will clear out all build artifacts including DBNet. By contrast, `tox -e clean` ignores DBNet.
- The proper way to now build the library should be done like the following:
    1. `tox -e lint`
    2. `git tag -a vX.X.X -m "Production release vX.X.X"` (use `vX.X.X.aX` for alpha builds, `.bX` for beta, `.devX`, and `.rcX`)
    3. `tox -e cleanall`
    4. `tox -e build`
    5. `tox -e publish`  <--- only for dev, alpha, beta, and release candidate builds of EasyOCR
    6. `tox -e publish -- --repository pypi`  <--- only for production releases of EasyOCR
- Building code via `setup.py` has been deprecated for some time now. As such, I have moved all configurations to `setup.cfg` in favor of using tox instructions which call `pytest`, `python -m build`, and `twine`. Each of these use `setup.cfg` for configuration. `pyproject.toml` would be better, however some tools have yet to add support for that standard, so we'll stick with `setup.cfg` for now.

Please let me know if you find anything unsavory with my work. I'll open a PR for this soon so @rkcosmos can review what has been done so far.

My best,
Jules

**Reply by JulianOrteil** [reply]
Created https://github.com/JaidedAI/EasyOCR/pull/850 to merge our changes related to automating maintainer tasks. I will keep it as a draft until you have taken a look at it yourself @ystoll. In the meantime, I will also be going on vacation until next week Tuesday. I may or may not work on this if I have free time while away.

**Reply by ystoll** [reply]
@JulianOrteil ,

Just got back from vacations.

I had a look at #850.
Everything seems fine to me.

Regards

Yannick

**Reply by JulianOrteil** [reply]
Sounds good and welcome back!

I have been absolutely swamped, so I haven't been able to really touch this since I opened 850. It is still looking like this may be the case for a couple more days, but let me know if there is anything I can help you with and I'll do my best to get back to you in a timely manner.

My best,
Jules

**Reply by JulianOrteil** [reply]
@ystoll If you have the opportunity, can you independently verify my work in the `metamation` branch if you're running Windows or Linux? Specifically, `tox -e build`.

On Linux:
- Run ``tox -e build``
- Create a new conda environment
- Change to a directory **outside** of EasyOCR
- ``pip install path/to/EasyOCR/dist/easyocr-<some version>.whl``
- Run ``import easyocr; reader = easyocr.Reader(['en'], detect_network = 'dbnet18'); reader.readtext("image.png")`` and let me know if you have errors

On Windows:
- Install VS2017 C++ Build Tools - installer: https://aka.ms/vs/15/release/vs_buildtools.exe
- Set a new ``CXX`` environment variable to MSVC compiler (typically ``C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\<VERSION>\bin\Hostx64\x64\cl.exe``)
- Run steps outlined in the Linux portion above

Note: If you have the CUDA toolkit installed (for example, ``C:\Program Files\NVIDIA GPU Computing Toolkit`` will exist on Windows), it may cause conflicts and raise errors. I had to uninstall my system CUDA toolkit, so you might have to as well.

**Reply by JulianOrteil** [reply]
I also do not have access to an Apple computer, so if you do and can test these changes on that platform, that'd be great. If anything breaks, I probably won't be able to help diagnose that platform though.

**Reply by ystoll** [reply]
@JulianOrteil 
Hi Julian, I will check your work on my platform and give you my feedback when that will be done. However, you should know that I am working on a Linux (Ubuntu 20.04) server as well. Hence, I do not have access to a Apple computer either. If critical, I can try to test the package installation procedure on Windows although I may be less at ease (generally I am trying to stick as much as possible to ubuntu for developing which for me is way much more convenient).

Concerning the test suits, how would you like that we split the workload ? Indeed, we should avoid spending time writing tests that will test the same functionalities...

Regards,
Yannick

**Reply by JulianOrteil** [reply]
@ystoll No worries about the platforms. I'll see if I can get a Linux KVM to run Windows with gpu passthrough so we can test Windows as we develop. I'm quite comfortable developing on both platforms, so we should be covered there. As for Apple, I can also try to get Linux KVM to run a Hackintosh, but no guarantees.

As for the workload, I agree. We'll be spending considerable amounts of time here. For how to split it up, I'm not quite sure. I don't think splitting it up by module (i.e. I write tests for ``moduleA.py`` and you write tests for ``moduleB.py``) will be very beneficial since most modules intermix. What do you think about writing a full application for testing instead of trying to test each individual method? Whatever we can't test with the application will then be tested manually, whether it be for a function/method or a special case with data. I've already configured tests to display coverage and fail if we don't achieve 100% coverage.

**Reply by JulianOrteil** [reply]
@ystoll Any thoughts on my proposal above?

**Reply by ystoll** [reply]
Hi @JulianOrteil,

sorry, I have been completely swamped myself working on my project ;-).
I have reviewed your P.R and everything looks good to be.

What do you mean by *writing a full application for testing* ? I am not sure to understand your idea here...

**Reply by JulianOrteil** [reply]
@ystoll No worries at all. I know how it gets!

My apologies for being vague and unclear. I was thinking of trying to kill two birds with one stone: an application that demonstrates EasyOCR that users can run while simultaneously being compatible with pytest to test the library. Whatever we can't test with the application is then explicitly tested; whether it be an obscure, buried method or an operation that only runs under specific circumstances.

**Reply by ystoll** [reply]
@JulianOrteil,

Hi Julian, I hope you are doing well. I would like to get back to you concerning your previous message:
* I think that it will be preferable to do things right the first time: writing a specific application for testing some parts of the code, and then testing the leftover parts with a classical approach seems, to me, not quite robust. I would rather starts by writing a full test suit, a little bit like what was done  for mmocr ([mmocr test suit](https://github.com/open-mmlab/mmocr/tree/main/tests)). FYI, I will start myself to write tests this afternoon, I hope that my first commit on that matter will be published tomorrow. 
* I think that the easiest way to keep us motivated, and here I speak mostly for myself would be to start iterating on the test suit asap.
Indeed, I perfectly aware that the two of us have many things to do on the side. But, I think that once we will get started, it will get easier to keep us going.
* To this end, I would like to open a private communication channel with you in order to send you messages which do not fit here. I communicate with my team via discord. If you are up for, we can use the same platform for the two of us to chat. You can ping me via my professional email available on my public Github account.

Have a nice day !

Yannick

**Reply by ystoll** [reply]
Hi @JulianOrteil , by the way, I just found that there is a testing method which already exist for constructing a test suit over a code base, which is exactly what we have to do here: the so-called  **golden master ** .  Maybe you already knew about this method - I didn't. Here are a few useful links about it:
* [blog post](https://tbenthompson.com/post/how_i_test/)
* [pytest-golden](https://github.com/oprypin/pytest-golden)
* [tutorial golden test](https://ro-che.info/articles/2017-12-04-golden-tests)

**Reply by JulianOrteil** [reply]
@ystoll No worries. I don't have any issue with writing tests as they should be. I just thought it would be a good idea to reduce duplicating how much code is tested multiple times by the same test methods.

Also, I sent you an email just now for setting up a private channel.

**Reply by JulianOrteil** [reply]
@ystoll I sent you a message on Discord a few days ago, just want to make sure it is you.

I've updated the "tests_suit" branch with my first attempt at using the `pytest-golden` framework. Please look through it and verify I've done it right as I haven't used this method of testing before. I don't want to write more tests until we've agreed on if this implementation I've attempted is correct.

Additionally, VSCode allows us to debug tests like we would normal Python code through their dedicated testing framework. However, there is an issue between `pdb` and `coverage` which prevents the former from working properly. If you happen to use VSCode and this debug mode, please add the following to your `.vscode/launch.json` file:
```
{
    "name": "Debug Unit Test",
    "type": "python",
    "request": "launch",
    "justMyCode": false,
    "program": "${file}",
    "purpose": ["debug-test"],
    "console": "integratedTerminal",
    "env": {
        "PYTEST_ADDOPTS": "--no-cov"
    },
}
```
This disables `coverage` during just debug testing, so running `tox` will still report coverage as expected.

**Comment by yuyang3478** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
